### ISSUE TITLE
Discussion: Simplifying the generation of release notes

### ISSUE THREAD

**Issue by hmaarrfk** [scikit-image/scikit-image#3404]
I'm summarizing some of the comments from the mailing list to open this topic up for discussion on github.
Please comment.

Currently the process of generating release notes is done at the time of a release.
Somebody needs to manually parse through all the commits and find all of the ones that are new for this release and summarize them in a one liner.

This is suboptimal for 2 reasons:

1.**Summarizing the PR:**  The person releasing scikit-image might not have been involved with that particular PR and thus needs to read through the whole thread to understand it.
2. **Finding the PRs:** This is non-trivial since it requires a manual search by hand.

On the mailing list, Mark (me) proposed that we ask people writing the PRs to write their own release notes. This would likely cause a conflict on every PR since every PR would be writing to the nearly the same line of the release notes file. The idea of asking new users to git to constantly rebase made this unsuitable for the open collaborative nature of scikit-image

Nathan mentioned tools like towncrier https://hypothesis.works/articles/continuous-releases/ that force a release on every PR. Much of the core team expressed the desire to decouple the PR process from the releasing processing and as such, this idea became unsuitable for scikit-image.

Juan suggested that the use of bots might be used to simplify the release process. However, he was conscious that the build process is already rather complicated and that adding complexity should be done with care. He wanted to ensure that we look for a solution that both addressed the problems relating to the build process while simplifying it.

**Comment by hmaarrfk** [comment]
I would like to add a suggestion:

At the end of the PR, the PR title should be formatted in a way suitable for inclusion in the release notes.

A demo of how the whole release notes could be generated programmatically can be found:
https://github.com/scikit-image/scikit-image/pull/3405

it might need a little bit of human intervention at the end to put similar PRs close together, but it would offload much of work (summary) to the time of authorship of the PR.

* Added complexity: low, we already use this script to generate the list of contributors and reviewers
* Friendliness to new contributors: They mostly have to do nothing. People in the scikit-image core team can change the title of their PRs already.
* Permanence: It generates a standard RST file, which means we can still archive it in the same way we used to before.

**Comment by jni** [comment]
@hmaarrfk 

> At the end of the PR, the PR title should be formatted in a way suitable for inclusion in the release notes.

I totally agree with this and it's the reason I added the "check PR title" tick box in the PR template after releasing 0.13.0. But it's hard to enforce and hard to remember to do when merging is done way at the bottom and the tick boxes are way at the top. A quick web search suggests that there is not currently a way to specify a merge commit template for core devs, though there is a bit of chatter about that, so it might happen at some point:

https://gitlab.com/gitlab-org/gitlab-ce/issues/40588
https://github.community/t5/How-to-use-Git-and-GitHub/How-to-change-the-default-quot-Squash-amp-Merge-quot-commit/td-p/9907

**Comment by hmaarrfk** [comment]
The really nice thing about the title of the PR is that it can be changed after the fact, by anybody. Anyway, just a suggestion. 

PR #3405 shows how we can autosort the changes too.

We can probably detect a keyword like "Backport" as well.
